sources/PPU/Background.cpp

```diff
@@ -2,27 +2,218 @@
+		int index = 0;
+		uint8_t reference = 0;
+		uint32_t color = 0;
+
+		tileData.raw = data;
+		palette = getPalette(tileData.palette);
+		// X horizontal
+		// Y vertical
+		graphicAddress = this->_tilesetAddress + (tileData.posY * NB_TILE_PER_ROW * this->_bpp * TILE_SIZE) + (tileData.posX * this->_bpp * TILE_SIZE);
+		for (int i = 0; i < this->_characterSize.y; i++) {
+			index = i * this->_characterSize.x;
+			if (tileData.verticalFlip)
+				index = (this->_characterSize.y - 1 - i) * this->_characterSize.x;
+			if (tileData.horizontalFlip)
+				index += this->_characterSize.x - 1;
+			for (int j = 0; j < this->_characterSize.x; j++) {
+				reference = getPixelReferenceFromTile(graphicAddress, index);
+				color = getRealColor(palette[reference]);
+				if (tileData.tilePriority == this->_priority) // reference 0 is considered as transparency
+					this->buffer[pos.x][pos.y] = (reference) ? color : 0;
+				index += (tileData.horizontalFlip) ? -1 : 1;
+				pos.x++;
 			}
+			pos.x -= this->_characterSize.x;
+			pos.y++;
+		}
+	}
+
+	std::vector<uint16_t> Background::getPalette(int nbPalette)
+	{
+		uint8_t nbColors = std::pow(2, this->_bpp);
+		uint16_t addr = nbPalette * this->_bpp * this->_bpp * 2; // 2 because it's 2 addr for 1 color
+		std::vector<uint16_t> palette(nbColors);
+
+		switch (this->_ppu.getBgMode()) {
+		case 0:
+			addr += (this->_bgNumber - 1) * (4 * 8) * 2;
+			break;
+		default:
+			break;
+		}
+
+		for (int i = 0; i < nbColors; i++) {
+			palette[i] = this->_cgram->read(addr);
+			palette[i] += this->_cgram->read(addr + 1) << 8U;
+			addr += 2;
+		}
+		return palette;
+	}
+
+	uint8_t Background::getPixelReferenceFromTile(uint16_t tileAddress, uint8_t pixelIndex)
+	{
+		uint8_t row = pixelIndex / this->_characterSize.x;
+		uint8_t column = pixelIndex % this->_characterSize.y;
+
+		if (row >= TILE_PIXEL_HEIGHT) {
+			tileAddress += 0x80 * this->_bpp;
+			row -= TILE_PIXEL_HEIGHT;
+		}
+		if (column >= TILE_PIXEL_WIDTH) {
+			tileAddress += 0x8 * this->_bpp;
+			column -= TILE_PIXEL_WIDTH;
+		}
+		// might not work with 8 bpp must check
+		tileAddress += 2 * row;
+
+		return this->getPixelReferenceFromTileRow(tileAddress, column);
+	}
+
+	uint8_t Background::getPixelReferenceFromTileRow(uint16_t tileAddress, uint8_t pixelIndex)
+	{
+		uint8_t highByte = this->_vram->read(tileAddress % VRAMSIZE);
+		uint8_t lowByte = this->_vram->read((tileAddress + 1) % VRAMSIZE);
+		uint8_t secondHighByte;
+		uint8_t secondLowByte;
+		uint16_t result = 0;
+		uint8_t shift = (TILE_PIXEL_WIDTH - 1U - pixelIndex);
+
+		switch (this->_bpp) {
+		case 8:
+			return highByte;
+		case 4:
+			secondHighByte =  this->_vram->read((tileAddress + 16) % VRAMSIZE);
+			secondLowByte = this->_vram->read((tileAddress + 17) % VRAMSIZE);
+			result = ((secondHighByte & (1U << shift)) | ((secondLowByte & (1U << shift)) << 1U));
+			result = (shift - 2 >= 0) ? result >> (shift - 2) : result << ((shift - 2) * -1);
+			__attribute__((fallthrough));
```

Comment:
You should use a macro for Fallthrough, using an attribute will allow the code to be compiled with something other than g++ easily.